core/src/com/ericc/the/game/mapGenerator/Generator.java

```diff
@@ -0,0 +1,147 @@
+
+        for (int i = x + roomX; i < x + roomWidth && i < this.width && i < x + width - 1; ++i) {
+            for (int j = y + roomY; j < y + roomHeight && j < this.height && j < y + height - 1; ++j) {
+                map[i][j] = Tile.generateRoomTile();
+
+                if (center.x == 0) {
+                    center.y++;
+                }
+            }
+
+            center.x++;
+        }
+
+        center.x = x + roomX + (center.x / 2);
+        center.y = y + roomY + (center.y / 2);
+
+        return center;
+    }
+
+    /**
+     * Indicates the rectangle of [x][y] to [x + width)[y + height)
+     * @param x indicates the x axis of the top right corner
+     * @param y same but indicates the y axis of the same corner
+     * @param width the width of a rectangle (from x till x + width (excluded))
+     * @param height the height of a rectangle (from y till y + height (excluded))
+     */
+    Coordinates<Integer> generateMap(int x, int y, int width, int height) {
+        if (width < this.maximalRoomSize || height < this.maximalRoomSize) {
+            return fillRoomReturnCenter(x, y, width, height);
+        }
+
+        boolean pickHeight = random.nextBoolean();
+        int slice = pickHeight ?
+                random.nextInt(1 + (height - 1) / 2) + height / 4 :
+                random.nextInt(1 + (width - 1) / 2) + width / 4;
+
+        Coordinates<Integer> firstRoom, secondRoom;
+
+        if (pickHeight) {
+            firstRoom = generateMap(x, y, width, slice);
+            secondRoom = generateMap(x, y + slice, width, height - slice);
+        } else {
+            firstRoom = generateMap(x, y, slice, height);
+            secondRoom = generateMap(x + slice, y, width - slice, height);
+        }
+
+        connectRooms(firstRoom, secondRoom);
+        return random.nextBoolean() ? firstRoom : secondRoom;
+    }
+
+    /**
+     * Generates the entire map.
+     */
+    void generateMap() {
+        generateMap(1, 1, width - 1, height - 1);
+    }
+
+    /**
+     * Given two centres of rooms, connects them making a new corridor.
+     *
+     * @param firstRoom  centre of a first room
+     * @param secondRoom centre of a second room
+     *                   <p>
+     *                   Disclaimer:
+     *                   Rooms can be given in any order. The function indicates
+     *                   the minimal (x-axis wise) and draws a corridor from x1 (minimal one)
+     *                   to x2 (maximal). Then does the same (from the position it is on at that time)
+     *                   but considers the y-axis this time - so starts from [x2][y1] and draws
+     *                   a corridor till it reaches [x2][y2]. (Assumptions for the explanation x1 <= x2
+     *                   y1 <= y2, the function look for that order on its own, though.)
+     */
+    void connectRooms(Coordinates<Integer> firstRoom, Coordinates<Integer> secondRoom) {
```

Comment:
to są jakieś wewnętrzne, implementacyjne sprawy -- chyba spokojnie mogą być private